testing/e2e/infra/provision.go

```diff
@@ -108,23 +106,45 @@ func (i *infra) Provision(ctx context.Context, tenantId, subscriptionId, applica
 		// that will change the loop variable capture to be the standard way loops work.
 		func(idx int) {
 			resEg.Go(func() error {
-				zone, err := clients.NewZone(ctx, subscriptionId, i.ResourceGroup, fmt.Sprintf("zone-%d-%s", idx, i.Suffix))
+				z, err := clients.NewZone(ctx, subscriptionId, i.ResourceGroup, fmt.Sprintf("Zone-%d-%s", idx, i.Suffix))
+				if err != nil {
+					return logger.Error(lgr, fmt.Errorf("creating Zone: %w", err))
+				}
+
+				<-kvDone
+
+				cert, err := ret.KeyVault.CreateCertificate(ctx, fmt.Sprintf("Zone-%d", idx), z.GetName(), []string{z.GetName()})
 				if err != nil {
-					return logger.Error(lgr, fmt.Errorf("creating zone: %w", err))
+					return logger.Error(lgr, fmt.Errorf("creating certificate: %w", err))
 				}
-				ret.Zones = append(ret.Zones, zone)
+
+				ret.Zones = append(ret.Zones, WithCert[Zone]{
+					Zone: z,
+					Cert: cert,
+				})
 				return nil
 			})
 		}(idx)
 	}
 	for idx := 0; idx < lenPrivateZones; idx++ {
```

Comment:
nit: why not have public and private zone creation happen, then block on kv creation, then block on kv creation for certs once? it requires another couple of for loops for private/public certs but allows more to happen while kv creation is going on in the background

Reply:
How would that allow more to happen while KV creation is blocking? The current pattern performs all possible work and only blocks once we require KV.

Reply:
you're so right -- missed the fact that the internal call is a goroutine